### PR TITLE
Clang++ provides its own exception types on Linux.

### PR DESCRIPTION
**Pull request by NorthSailor**
Without disabling the existing definitions, clang will refuse to build libobjc2 on Ubuntu Linux 18.04. With this patch, the library compiles without problems.

### CODE REVIEW

**Comment by davidchisnall** [comment]
These changes don't build and it's not clear what they're actually trying to do.

**Comment by NorthSailor** [comment]
Note that I am using Clang 6, while Travis has Clang 5.

**Comment by davidchisnall** [comment]
A fix that breaks existing systems won't be committed.  A fix that does not explain the problem that it's trying to solve will not be committed.

I believe that this is related to namespace pollution in recent versions of the libstdc++ headers, which include Itanium C++ ABI functions when including C++ standard library headers.

**Comment by davidchisnall** [comment]
Okay, so it appears that you are seeing the same issue as #66.  Does #67 fix it?  If so, then we'll merg that.  If not, then please see the discussion in #66 for the correct fix: we need to add a check that tests whether we need to add a `nothrow` qualifier on the function.  We also want to name the check correctly.  The [standard](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html) does not put a `nothrow` qualifier on this definition, so it is a work around for a platform toolchain not providing definitions that are compatible with the standard ones.

**Comment by NorthSailor** [comment]
Yeah, sorry I missed the issue. Working on a platform test right now.

File: CMakeLists.txt [hunk 1]
```diff
@@ -367,9 +369,27 @@ if (TESTS)
...
+	extern \"C\" {
+	__attribute__((weak))
+	void *__cxa_allocate_exception(size_t thrown_size);
+	}
+	#include <exception>
+	int main() { return 0; }" CXA_ALLOCATE_EXCEPTION_CAN_THROW_COMPILES)
+
+if (NOT CXA_ALLOCATE_EXCEPTION_CAN_THROW_COMPILES)
+	add_definitions(-DCXA_ALLOCATE_EXCEPTION_IS_NOEXCEPT)
+endif ()
```
[the remark this answers is not in the document]
Comment (NorthSailor) [reply CMakeLists.txt]: Okay, I will invert the check.

File: objcxx_eh.cc [hunk 2]
```diff
@@ -15,7 +15,7 @@ namespace __cxxabiv1
 
 using __cxxabiv1::__class_type_info;
 
-#ifndef USE_STD_EXCEPTION_TYPES
+#ifndef TYPE_INFO_IS_EXPOSED
```
[the remark this answers is not in the document]
Comment (NorthSailor) [reply objcxx_eh.cc]: image
I can remove it from this merge request, since it probably deserves an issue on its own.
No test seems to fail even after removing it though.

File: objcxx_eh.h [hunk 3]
```diff
@@ -10,10 +10,12 @@ extern "C" {
  * __cxa__allocate_exception(). Since the standard version is defined with
  * _GLIBCXX_NOTHROW, clang gives a type mismatch error.
  */
-#if !__cplusplus || !USE_STD_EXCEPTION_TYPES
 __attribute__((weak))
-void *__cxa_allocate_exception(size_t thrown_size);
+void *__cxa_allocate_exception(size_t thrown_size)
+#if defined __cplusplus && defined CXA_ALLOCATE_EXCEPTION_IS_NOEXCEPT
```
[the remark this answers is not in the document]
Comment (NorthSailor) [reply objcxx_eh.h]: I'm sorry but I'm not 100% sure what you mean here. It is a single -D flag from line 383 of `CMakeLists.txt`. The C++ check is necessary without it `eh_personality.c` fails to compile since it gets a `noexcept` in the C code. But I can make it a bit more readable by typing the function declaration twice, if you would prefer that, like so:

```
#if defined __cplusplus && defined CXA_ALLOCATE_EXCEPTION_IS_NOEXCEPT
__attribute__((weak))
void  *__cxa_allocate_exception(size_t thrown_size) noexcept;
#else
__allocate((weak))
void *__cxa_allocate_exception(size_t thrown_size);
#endif
```

[the remark this answers is not in the document]
Comment (davidchisnall) [reply objcxx_eh.h]: Flag should be either `-DFOO=` or `-DFOO=noexcept`, and then use `FOO` in the code (where `FOO` is a more sensible name, such as `CXA_ALLOCATE_EXCEPTION_NOEXCEPT`)

File: CMakeLists.txt [hunk 4]
```diff
@@ -162,7 +164,6 @@ set_source_files_properties(
 # C++ Runtime interaction
 #
 
-
```
Comment (davidchisnall) [inline CMakeLists.txt]: Please remove whitespace change.

File: CMakeLists.txt [hunk 5]
```diff
@@ -368,3 +369,26 @@ if (TESTS)
...
+else ()
+	add_definitions(-DCXA_ALLOCATE_EXCEPTION_SPECIFIER=)
+endif ()
+
+CHECK_CXX_SOURCE_COMPILES("
+	#include <exception>
+	namespace std { class type_info { }; }
+	int main() { return 0; }" TYPE_INFO_REDEFINITION_COMPILES)
+
+if (NOT TYPE_INFO_REDEFINITION_COMPILES)
```
Comment (davidchisnall) [inline CMakeLists.txt]: Please put this in a separate PR once you've explained what problem it is solving.

File: objcxx_eh.h [hunk 6]
```diff
@@ -5,8 +5,18 @@ extern "C" {
  * Allocates a C++ exception.  This function is part of the Itanium C++ ABI and
  * is provided externally.
  */
+/*
+ * Note: Recent versions of libstdc++ already provide a prototype for
```
Comment (davidchisnall) [inline objcxx_eh.h]: s/libstdc++/libsupc++/

File: objcxx_eh.h [hunk 7]
```diff
@@ -5,8 +5,18 @@ extern "C" {
...
+ * __cxa__allocate_exception(). Since the standard version is defined with
```
Comment (davidchisnall) [inline objcxx_eh.h]: s/standard/libsupc++/
